### PR TITLE
enable random PMA memory usage in corev-dv tests

### PR DESCRIPTION
**Pull request by strichmo**
# corev-dv random PMA

Merges in @silabs-hfegran's work to enable:
- Multiple PMA configurations to be used in corev-dv tests
- corev-dv instruction libraries to use multiple regions across different PMAs

# All generated files to results
'
The generated directories for test_programs has been overhauled to move all of the generated files into this type of directory:

```
<sim>_results/<cfg>/<test>/<run_index>/test_program/
```

The BSP also is built into the results area on a per-config basis:
```
<sim>_results/<cfg>/bsp
```

*Going forward we should not be generating temporary files in the core-v-verif source tree.*

# Regression support for configurations

Regression support for configurations and tests was greatly enhanced.  release check and ci check test lists were updated accordingly.

# Compatibility
Xcelium, Questa and DSIM should pass in E40X and E40P with these updates.
VCS and Riviera were updated to follow changes made in the other simulators, however I have no way to test that.

# Known Issues
- The loading of the ELF file into OVPSIM for certain PMA configurations can take quite a long time, especially in DSIM.  The linker generation will need to be smarter to better allocate smaller regions for memory actually used.
- Directed tests do not have a way to use the random PMA regions as these linker scripts are generated by corev-dv.

### CODE REVIEW

File: cv32e40x/tb/uvmt/uvmt_cv32e40x_ldgen_tb.sv
Comment (MikeOpenHWGroup) [inline cv32e40x/tb/uvmt/uvmt_cv32e40x_ldgen_tb.sv:1]: Why do we need two of these?   There is already `cv32e40x/tb/ldgen/ldgen_tb.sv`.  I am missing something.

File: cv32e40x/tests/programs/corev-dv/corev_rand_arithmetic_base_test/.gitignore
Comment (MikeOpenHWGroup) [inline cv32e40x/tests/programs/corev-dv/corev_rand_arithmetic_base_test/.gitignore:1]: Thanks for deleting these.  They are redundant and lead to maintenance issues in the future.

File: cv32e40x/tests/programs/custom/cv32e40x_csr_access_test/cv32e40x_csr_access_test.S
Comment (MikeOpenHWGroup) [inline cv32e40x/tests/programs/custom/cv32e40x_csr_access_test/cv32e40x_csr_access_test.S:1602]: Nice.

File: lib/mem_region_gen/pma_adapted_mem_region_gen.sv
Comment (MikeOpenHWGroup) [inline lib/mem_region_gen/pma_adapted_mem_region_gen.sv:18]: You are gonna hate me for this one.  No SPDX leader.

File: mk/Common.mk
Comment (MikeOpenHWGroup) [inline mk/Common.mk:251]: I'll accept this (because its the right thing to do), but will need to check close to see if it causes problems for the CORE testbench.   Leave it to me.

**Review by MikeOpenHWGroup — CHANGES_REQUESTED** [review]
Whew!   What a monster.   I've left a few sofball questions/comments in this review, but would appreciate an in-person discussion before merging it in.    FWIW, it passes a local regression with DSIM.

File: cv32e40x/tb/uvmt/uvmt_cv32e40x_ldgen_tb.sv
Comment (silabs-hfegran) [reply cv32e40x/tb/uvmt/uvmt_cv32e40x_ldgen_tb.sv:1]: @MikeOpenHWGroup, Sorry, you are completely right, this file is a remnant from when I ported the standalone ldgen to use UVM. This file should no longer be used, and the ldgen_tb.sv is used if standalone generation of linker scripts is desired.

File: cv32e40x/tb/ldgen/ldgen_tb.sv
Comment (silabs-hfegran) [reply cv32e40x/tb/ldgen/ldgen_tb.sv:13]: The standalone tb is used if we want to generate valid linker scripts for e.g. directed tests. At one point I used this target to generate an automated pma setup for directed tests with an extra invocation of xcelium, however it was decided to remove it from the current makefile due to the added complexity and change of build flow this would incur. 

Furthermore this target tb is very useful in debugging of the pma-setup generation - without invoking the complete core-v-dv flow

File: cv32e40x/bsp/link_corev-dv.ld
Comment (silabs-hfegran) [reply cv32e40x/bsp/link_corev-dv.ld:1]: Agreed - this part is a straight copy from the default BSP linker files with minimum amount of necessary changes.

File: cv32e40x/tests/programs/custom/generic_exception_test/generic_exception_test.S
Comment (silabs-hfegran) [inline cv32e40x/tests/programs/custom/generic_exception_test/generic_exception_test.S:214]: Not part of this review per se, but maybe a more informative label would be useful here?

File: lib/mem_region_gen/pma_adapted_mem_region_gen.sv
Comment (silabs-hfegran) [inline lib/mem_region_gen/pma_adapted_mem_region_gen.sv:58]: Unless needed for compatibility reasons, I prefer the proess_stack to be a bounded queue of maximum two elements [$:1] - Leaving it unbounded could mask potential bugs and make them much harder to find. (Been there, debugged that....). This queue should never exceed 2 elements, so if the queue is not bounded there should be an assertion checking that this is never exceeded.

File: cv32e40x/env/corev-dv/cv32e40x_asm_program_gen.sv
Comment (silabs-hfegran) [inline cv32e40x/env/corev-dv/cv32e40x_asm_program_gen.sv:38]: Remove this commented out block

File: cv32e40x/env/corev-dv/cv32e40x_asm_program_gen.sv
Comment (silabs-hfegran) [reply cv32e40x/env/corev-dv/cv32e40x_asm_program_gen.sv:81]: Agree that this looks strange - This is because this function is copied directly from riscv-dv and added necessary features to preserve as much commonality as possible.

File: cv32e40x/env/corev-dv/cv32e40x_asm_program_gen.sv
Comment (silabs-hfegran) [inline cv32e40x/env/corev-dv/cv32e40x_asm_program_gen.sv:83]: We want to turn these cfg.pmp_cfg-blocks back on when we start up PMP-testing - Needs further testing in combination with PMA

File: cv32e40x/env/corev-dv/cv32e40x_asm_program_gen.sv
Comment (silabs-hfegran) [inline cv32e40x/env/corev-dv/cv32e40x_asm_program_gen.sv:378]: This can be removed

File: cv32e40x/env/corev-dv/cv32e40x_asm_program_gen.sv
Comment (silabs-hfegran) [reply cv32e40x/env/corev-dv/cv32e40x_asm_program_gen.sv:387]: No longer used - was used in a previous (duplicate) nmi-handler that caused a lot of issues for the linker, should also be removed

File: cv32e40x/env/corev-dv/cv32e40x_instr_base_test.sv
Comment (silabs-hfegran) [reply cv32e40x/env/corev-dv/cv32e40x_instr_base_test.sv:37]: Yup, this is where the linker scripts are generated in the corev-dv flow, the ldgen tb is used for standalone generation as mentioned in my other comment

File: cv32e40x/env/corev-dv/cv32e40x_instr_base_test.sv
Comment (silabs-hfegran) [inline cv32e40x/env/corev-dv/cv32e40x_instr_base_test.sv:46]: Remove comments

File: cv32e40x/env/corev-dv/ldgen/cv32e40x_ldgen.sv
Comment (silabs-hfegran) [inline cv32e40x/env/corev-dv/ldgen/cv32e40x_ldgen.sv:1]: @MikeOpenHWGroup:  This is the place where the actual generation of the linker scripts happen - both by constructor in the corev-dv build phase and in ldgen_tb

File: cv32e40x/env/corev-dv/ldgen/cv32e40x_ldgen.sv
Comment (silabs-hfegran) [inline cv32e40x/env/corev-dv/ldgen/cv32e40x_ldgen.sv:275]: Remove commented code

File: cv32e40x/env/corev-dv/ldgen/cv32e40x_ldgen.sv
Comment (silabs-hfegran) [inline cv32e40x/env/corev-dv/ldgen/cv32e40x_ldgen.sv:402]: Remove commented code

File: cv32e40x/env/corev-dv/cv32e40x_asm_program_gen.sv
Comment (silabs-hfegran) [inline cv32e40x/env/corev-dv/cv32e40x_asm_program_gen.sv:38]: remove commented code

File: cv32e40x/env/corev-dv/ldgen/cv32e40x_ldgen.sv
Comment (silabs-hfegran) [inline cv32e40x/env/corev-dv/ldgen/cv32e40x_ldgen.sv:494]: Remove commented code

File: cv32e40x/env/corev-dv/ldgen/cv32e40x_ldgen.sv
Comment (silabs-hfegran) [inline cv32e40x/env/corev-dv/ldgen/cv32e40x_ldgen.sv:504]: remove commented code

File: cv32e40x/sim/Common.mk
Comment (silabs-hfegran) [reply cv32e40x/sim/Common.mk:23]: I do not think that would pose any significant problems

File: cv32e40x/tests/programs/custom/debug_test_trigger/trigger_code.S
Comment (silabs-hfegran) [inline cv32e40x/tests/programs/custom/debug_test_trigger/trigger_code.S:281]: could these magic numbers be put into some constant like the previous CV_VP_* constants?

File: lib/mem_region_gen/pma_adapted_mem_region_gen.sv
Comment (silabs-hfegran) [inline lib/mem_region_gen/pma_adapted_mem_region_gen.sv:86]: Add comment to this function

File: mk/uvmt/uvmt.mk
Comment (silabs-hfegran) [inline mk/uvmt/uvmt.mk:85]: Great! These common SIM_* variables do wonders for maintainability

File: cv32e40x/tests/uvmt/test-programs/syscalls.c
Comment (silabs-hfegran) [inline cv32e40x/tests/uvmt/test-programs/syscalls.c:28]: Should we define these externally as well?

**Review by silabs-hfegran — CHANGES_REQUESTED** [review]
I will submit a PR for most of the "Remove commented code"-parts I mentioned in the relevant sections - There are some other things I commented on in the applicable sections that I would like some feedback on before approving the review.

File: cv32e40x/bsp/link_corev-dv.ld
Comment (strichmo) [reply cv32e40x/bsp/link_corev-dv.ld:1]: Fixed.

File: cv32e40x/env/corev-dv/cv32e40x_asm_program_gen.sv
Comment (strichmo) [reply cv32e40x/env/corev-dv/cv32e40x_asm_program_gen.sv:71]: Chased this down...I say we should leave these as is.  The actual prototype of the method from the riscv_asm_program_gen.sv is:

  function void gen_signature_handshake(ref string instr[$],
                                        input signature_type_t signature_type,
                                        core_status_t core_status = INITIALIZED,
                                        test_result_t test_result = TEST_FAIL,
                                        privileged_reg_t csr = MSCRATCH,
                                        string addr_label = "");

File: cv32e40x/tb/ldgen/ldgen_tb.sv
Comment (strichmo) [reply cv32e40x/tb/ldgen/ldgen_tb.sv:13]: I believe the intent has been adequately communicated so closing this thread.

File: lib/mem_region_gen/pma_adapted_mem_region_gen.sv
Comment (strichmo) [reply lib/mem_region_gen/pma_adapted_mem_region_gen.sv:18]: Fixed.

File: cv32e40x/tests/programs/custom/generic_exception_test/generic_exception_test.S
Comment (strichmo) [reply cv32e40x/tests/programs/custom/generic_exception_test/generic_exception_test.S:214]: Noted that general exception test cleanup should be done for E40X and future cores.

**Review by MikeOpenHWGroup — COMMENTED** [review]
I am ready to approve this as soon as @strichmo and @silabs-halfdan give their thumbs up.

**Comment by strichmo** [comment]
It should be @silabs-hfegran instead of @silabs-halfdan !

**Review by silabs-hfegran — APPROVED** [review]
> It should be @silabs-hfegran instead of @silabs-halfdan !

No worries @MikeOpenHWGroup !

looks A-OK to me, thanks for the changes @strichmo

File: cv32e40x/tests/programs/custom/debug_test_trigger/trigger_code.S
Comment (strichmo) [reply cv32e40x/tests/programs/custom/debug_test_trigger/trigger_code.S:281]: This is actually an encoded VP register, but I will add macros to at least better document the test intention.

File: lib/mem_region_gen/pma_adapted_mem_region_gen.sv
Comment (strichmo) [reply lib/mem_region_gen/pma_adapted_mem_region_gen.sv:91]: Can't ignore cleaning up grammar! :-)  This is fixed.

**Comment by strichmo** [comment]
@MikeOpenHWGroup @silabs-hfegran Thanks very much for your reviews.  This PR branch should be completely up-to-date with all comments.

I went ahead and merged the latest updates from cv32e40x/dev so that I could update to the latest RTL hash as well.

**Review by MikeOpenHWGroup — APPROVED** [review]

**Pull request merged**
